// react-spring-three/index.js
import * as THREE from 'three'
import { applyProps } from 'react-three-fiber'
import {
  Globals,
  interpolate,
  animated,
  Interpolation,
  config,
  useChain,
  useSpring,
  useSprings,
  useTrail,
  useTransition,
  createInterpolation,
  colorNames,
} from 'react-spring/cookbook'

// Extend animated with all the available THREE elements
const extendedAnimated = Object.keys(THREE).reduce((acc, element) => {
  const key = element[0].toLowerCase() + element.substring(1)
  acc[key] = animated(key)
  return acc
}, animated)

// Set default native-element
Globals.injectDefaultElement('group')
// Use default interpolation (which includes numbers, strings, colors)
Globals.injectInterpolation(createInterpolation)
// Inject color names, so that it will be able to deal with things like "peachpuff"
Globals.injectColorNames(colorNames)
// This is how we teach react-spring to set props "natively", the api is (instance, props) => { ... }
Globals.injectApplyAnimatedValues(applyProps)

export {
  config,
  extendedAnimated as animated,
  interpolate,
  Globals,
  useSpring,
  useTrail,
  useTransition,
  useChain,
  useSprings,
  Interpolation,
}
